### PR TITLE
SplAdder Pipeline (resolves #145)

### LINKED ISSUE #145: Write Spladder/Alt-Splice Pipeline
- variant calling ~ 1h / single thread / 1G RAM
- AS calling ~ 2.5 h / single thread / 5G RAM
- QC run ~10min / single thread / < 1G RAM

I would suggest to lump variant calling and QC together. 

As input data, I have currently used the hg38 Reference genome downloaded from ensembl and as annotation the Gencode V23 comprehensive. Let me know when you need more information. 

Full list of dependencies:

Variant calling step:
===============
Requirements:
* samtools - version 1.3

Pre-requisits:
* index genome.fa with: samtools faidx <genome.fa>
(this needs to be done only once and the genome.fa and the index can be provided as input files)

Call:
* samtools-1.3 mpileup -f [genome.fa] -l [position_file] -v [alignment.bam] -t DP,SP,INFO/AD,INFO/ADF,INFO/ADR,INFO/DPR,SP > [output.vcf]

Resources:
* time: ~ 1h
* memory: ~ 300MB
* disk space: only for result file ~ 10MB

Source:
http://www.htslib.org/download/

Alternative splicing step:
===================
Requirements:
* python 2.7
* pysam
* h5py
* numpy
* scipy

Pre-requisits:
* needs genome.fa
* needs annotation.gtf

Call:
* python spladder.py --insert_ir=y --insert_es=y --insert_ni=y --remove_se=n --validate_sg=n -b [alignment.bam] -o [outdir] -a [annotation.gtf] -v y -c 3 -M merge_graphs -T y -n [readlength] -P y -p n --sparse_bam y
(we can set the read length to constant 50 for now - this will only influence some filter criteria and should be fine - unless you have a large fraction of 32bp reads in your sets, which is unlikely)

Resources:
* time: ~ 2-3h
* memory: ~ 4 GB
* single CPU
* disk space: ~ 1GB	

Source:
https://github.com/ratschlab/spladder

QC tool:
=======
Requirements:
* python 2.7
* pysam
* pandas
* scipy
* numpy

Pre-requisits:
* needs pre-processed annotation table - will be provided by us

Call:
* checkBias_2.0.py -o [outfile-prefix] -n [alignment.bam] -v -a [annotation.gtf] -m [annotation.preprocessed.m53]

Resources:
* time: ~20min
* memory: < 1GB
* disk space: negligible

Source:
* https://github.com/KjongLehmann/m53

### PR DESCRIPTION
**Pull request by jvivian**
Sphinx docstrings
Common lib functionality

### CODE REVIEW

File: src/toil_scripts/spladder_pipeline/spladder_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+    if args.output_s3_dir:
+        assert args.output_s3_dir.startswith('s3://'), 'Wrong format for output s3 directory'
+    # Program checks
+    for program in ['curl', 'docker']:
+        assert which(program), 'Program "{}" must be installed on every node.'.format(program)
+
+    Job.Runner.startToil(Job.wrapJobFn(parse_input_samples, args), args)
+
+
+if __name__ == '__main__':
```
Comment (hannes-ucsc) [inline src/toil_scripts/spladder_pipeline/spladder_pipeline.py]: We should remove this so people aren't even tempted to run this as a script.

File: src/toil_scripts/spladder_pipeline/launch_spladder_mesos.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+aws:us-west-2:spladder-run-1 \
+--config /home/mesosbox/shared/config.txt \
+--retryCount 1 \
+--ssec /home/mesosbox/shared/master.key \
+--output-s3-dir s3://bucket/dir \
+--sseKey=/home/mesosbox/shared/master.key \
+--batchSystem="mesos" \
+--mesosMaster mesos-master:5050 \
+--workDir=/var/lib/toil \
+# --restart
```
Comment (hannes-ucsc) [inline src/toil_scripts/spladder_pipeline/launch_spladder_mesos.sh]: NL missing at EOL

**Comment by hannes-ucsc** [comment]
@jvivian, done with review.

**Comment by jvivian** [comment]
@hannes-ucsc — Squashed and rebased. Commit lists review considerations.

**Pull request merged**